File: leetcode/101-200/leetcode142.cpp
#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x)
        : val(x)
        , next(NULL)
    {
    }
};
const int OVERANGE = 100005;
int pos;
class Solution {
public:
    ListNode* detectCycle(ListNode* head)
    {
        ListNode *fast, *slow, *ans;
        fast = slow = ans = head;
        if (head == nullptr || head->next == nullptr)
            return nullptr;
        while (fast->next != nullptr && fast->next->next != nullptr) {
            fast = fast->next->next;
            slow = slow->next;
            if (fast == slow) {
                while (slow != ans) {
                    ans = ans->next;
                    slow = slow->next;
                }
                return ans;
            }
        }
        return nullptr;
    }
};
int main(void)
{
    return 0;
}

File: leetcode/100-999/leetcode142.py
# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return None
        slow = fast = ans = head
        while True:
            if not fast.next or not fast.next.next:
                return None
            fast = fast.next.next
            slow = slow.next

            if slow == fast:
                while ans != slow:
                    ans = ans.next
                    slow = slow.next
                break
        
        return slow
        